### PR TITLE
Use the current firewall token if it matches the context

### PR DESCRIPTION
**Pull request by aschempp**
If the user is logged in in the current request (e.g. through *rememberme*), the session key does not exist, but the user exists in the firewall. We must always use/respect the current firewall if it matches the requested scope.

### CODE REVIEW

File: core-bundle/src/Resources/config/services.yml [hunk 1]
```diff
@@ -510,6 +510,9 @@ services:
     contao.security.token_checker:
         class: Contao\CoreBundle\Security\Authentication\Token\TokenChecker
         arguments:
+            - '@request_stack'
+            - '@security.firewall.map'
+            - '@security.token_storage'
```
Comment (leofeyer) [inline core-bundle/src/Resources/config/services.yml:515]: Just curious: Why did you not add the new arguments at the end?

Comment (aschempp) [reply core-bundle/src/Resources/config/services.yml:515]: 🤷‍♂️ Just because the new ones are kinda more important, the existing ones are not used anymore if the new ones „match“.

File: core-bundle/src/Security/Authentication/Token/TokenChecker.php [hunk 2]
```diff
@@ -96,11 +119,11 @@ public function isPreviewMode(): bool
 
     private function getToken(string $sessionKey): ?TokenInterface
     {
-        if (!$this->session->isStarted() || !$this->session->has($sessionKey)) {
-            return null;
-        }
+        $token = $this->getTokenFromStorage(substr($sessionKey, 10));
```
Comment (leofeyer) [inline core-bundle/src/Security/Authentication/Token/TokenChecker.php]: Why did you add `substr($sessionKey, 10)` here? Should we not throw an `InvalidArgumentException` if the session key is too long?

Comment (aschempp) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php]: Its not about the length, the session key contains more that the necessary firewall context.

File: core-bundle/src/Resources/config/services.yml (hunk 1 above)
Comment (leofeyer) [reply core-bundle/src/Resources/config/services.yml:515]: Absolutely reasonable from my point of view 😄

File: core-bundle/src/Security/Authentication/Token/TokenChecker.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php]: I see. We should document this then I guess. 😆

Comment (aschempp) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php]: We should actually never have added the session kex constants but use the scopes…

Comment (leofeyer) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php]: Is it too late to change it?

Comment (aschempp) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php]: well we could remove the use of it, but the constants _could_ be used by people? Not sure if they are API?

Comment (leofeyer) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php]: We don't need to remove them. We can deprecate them and stop using them.

**Comment by leofeyer** [comment]
@aschempp If you fix the unit tests, I will merge this PR.

File: core-bundle/src/Security/Authentication/Token/TokenChecker.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php]: So, does it make sense to hardcode this to `_security_` then?

Comment (aschempp) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php]: yes, it's hardcoded in Symfony as well: https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/Security/Http/Firewall/ContextListener.php#L64

File: core-bundle/src/Security/Authentication/Token/TokenChecker.php [hunk 3]
```diff
@@ -112,4 +138,30 @@ private function getToken(string $sessionKey): ?TokenInterface
...
+        $config = $this->firewallMap->getFirewallConfig($request);
+
+        if (!$config instanceof FirewallConfig || $config->getContext() !== $context) {
+            return null;
+        }
+
+        return $this->tokenStorage->getToken();
+    }
+
+    private function getTokenFromSession(string $sessionKey)
```
Comment (Toflar) [inline core-bundle/src/Security/Authentication/Token/TokenChecker.php]: Return typehint or phpdoc missing

Comment (aschempp) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php]: Fixed in 4d0b52bef783333027ec22e292d23ce52dbf0d76

**Comment by leofeyer** [comment]
Thank you @aschempp.
